onnxruntime/core/mlas/lib/sqnbitgemm_kernel_neon.cpp

```diff
@@ -0,0 +1,502 @@
+    // a0_2 a1_2 a3_2 a3_2
+    a2 = vreinterpretq_f32_f64(vzip1q_f64(vreinterpretq_f64_f32(b1), vreinterpretq_f64_f32(b3)));
+    // a0_3 a1_3 a2_3 a3_3
+    a3 = vreinterpretq_f32_f64(vzip2q_f64(vreinterpretq_f64_f32(b1), vreinterpretq_f64_f32(b3)));
+
+    return vaddq_f32(vaddq_f32(a0, a1), vaddq_f32(a2, a3));
+}
+
+template <size_t BlkBitWidth, size_t BlkLen, size_t NCols>
+MLAS_FORCEINLINE void
+ComputeDotProducts(
+    const float* ARowPtr,
+    const uint8_t* QuantBDataColPtr,
+    const float* QuantBScaleColPtr,
+    const uint8_t* QuantBZeroPointColPtr,
+    float* SumPtr,
+    size_t CountK,
+    size_t StrideQuantBData,
+    size_t StrideQuantBScale,
+    size_t StrideQuantBZeroPoint,
+    const float* BiasPtr
+)
+{
+    static_assert(NCols == 1 || NCols == 4, "NCols must be 1 or 4");
+
+    const uint8x8_t LowMask = vdup_n_u8(0x0F);
+
+    float32x4_t acc[NCols]{};
+
+    const uint8_t* QuantBData = QuantBDataColPtr;
+    const float* QuantBScale = QuantBScaleColPtr;
+    size_t QuantBZeroPointIdx = 0;  // track half byte increments with this idx instead of a pointer
+
+    for (size_t k = 0; k < CountK; k += BlkLen) {
+        const size_t k_blk_len = std::min(CountK - k, BlkLen);
+
+        const uint8_t* b_data[NCols];
+        UnrolledLoop<NCols>(
+            [&](size_t i) { b_data[i] = QuantBData + i * StrideQuantBData; }
+        );
+
+        float scale[NCols];
+        UnrolledLoop<NCols>(
+            [&](size_t i) { scale[i] = QuantBScale[i * StrideQuantBScale]; }
+        );
+
+        uint8_t zp[NCols];
+        if (QuantBZeroPointColPtr != nullptr) {
+            UnrolledLoop<NCols>([&](size_t i) {
+                uint8_t zp_packed =
+                    QuantBZeroPointColPtr[i * StrideQuantBZeroPoint + QuantBZeroPointIdx / 2];
+                zp[i] = ((QuantBZeroPointIdx & 1) == 1) ? (zp_packed >> 4) : (zp_packed & 0x0F);
+            });
+        } else {
+            UnrolledLoop<NCols>([&](size_t i) {
+                zp[i] = 8;
+            });
+        }
+
+        constexpr size_t SubBlkLen = 16;  // number of block elements to process in one iteration
+
+        for (size_t k_idx_in_blk = 0; k_idx_in_blk < k_blk_len; k_idx_in_blk += SubBlkLen) {
+            // load A row vector elements
+
+            // load `SubBlkLen` elements from A padding with 0's if there aren't enough
+            float a_segment[SubBlkLen];
+            {
+                const size_t k_subblk_len = std::min(k_blk_len - k_idx_in_blk, SubBlkLen);
+                const float* a_begin = ARowPtr + k + k_idx_in_blk;
+                std::copy(a_begin, a_begin + k_subblk_len, a_segment);
+                std::fill(a_segment + k_subblk_len, a_segment + SubBlkLen, 0.0f);
+            }
+
+            // `SubBlkLen` floats of A
+            float32x4_t av[4];
+            UnrolledLoop<4>([&](size_t i) { av[i] = vld1q_f32(a_segment + i * 4); });
```

Comment:
combine this with intrinsic load

Reply:
[https://arm-software.github.io/acle/neon_intrinsics/advsimd.html#load-1](https://arm-software.github.io/acle/neon_intrinsics/advsimd.html#load-1)